library/ns1_zone.py

```diff
@@ -110,173 +140,323 @@
-                               supports_check_mode=True,
-                               mutually_exclusive=self.mutually_exclusive)
-        self.exec_module()
-
-    def api_params(self):
-        params = dict(
-            (key, self.module.params.get(key))
-            for key in ZONE_KEYS
-            if self.module.params.get(key) is not None
+        NS1ModuleBase.__init__(
+            self,
+            self.module_arg_spec,
+            supports_check_mode=True,
+            mutually_exclusive=self.mutually_exclusive,
         )
-        return params
 
-    def get_zone(self):
+    def sanitize_params(self, params):
+        """Removes all Ansible module parameters from dict that have no value
+        or are listed in SANITIZED_PARAMS
+
+        :param params: Ansible module parameters
+        :type params: dict
+        :return: Sanitized dict of params
+        :rtype: dict
+        """
+        sanitized = {}
+        for k, v in params.items():
+            if isinstance(v, dict):
+                v = self.sanitize_params(v)
+            if v is not None and k not in SANITIZED_PARAMS:
+                sanitized[k] = v
+        return sanitized
+
+    def get_zone(self, name):
+        """Retrieves a zone from NS1. If no name is given or zone does not
+        exist, will return None.
+
+        :param name: Name of the zone to retrieve
+        :type name: str, optional
+        :return: zone object returned by NS1
+        :rtype: ns1.zones.Zone
+        """
         zone = None
-        try:
-            zone = self.ns1.loadZone(self.module.params.get('name'))
-        except ResourceException as re:
-            if re.response.code != 404:
-                module.fail_json(
-                    msg="error code %s - %s " % (re.response.code, re.message)
-                )
-                zone = None
-        return zone
-
-    def update(self, zone):
-        changed = False
-        args = {}
-
-        for key in ZONE_KEYS:
-            if (
-                    self.module.params.get(key) is not None and
-                    (
-                        not zone.data or
-                        key not in zone.data or
-                        self.module.params.get(key) != zone.data[key]
+        if name:
+            try:
+                zone = self.ns1.loadZone(name)
+            except ResourceException as re:
+                if re.response.code != 404:
+                    self.module.fail_json(
+                        msg="error code %s - %s "
+                        % (re.response.code, re.message)
                     )
-            ):
-                changed = True
-                args[key] = self.module.params.get(key)
+                    zone = None
+        return zone
 
-        if self.module.check_mode:
-            # check mode short circuit before update
-            self.module.exit_json(changed=changed)
+    def compare_params(self, have, want):
```

Comment:
should we put this somewhere else? I'm guessing this is going to be used by other resources once those are refactored? Or were you going to wait until we do that to pull this out somewhere common?

Reply:
You called it, was planning on waiting to pull this out until we refactor `ns1_record` module.